src/corehost/corehost.cpp

```diff
@@ -338,5 +353,39 @@ int main(const int argc, const pal::char_t* argv[])
         trace::info(_X("}"));
     }
 
-    return run(argc, argv);
+#if defined(_WIN32) && defined(FEATURE_APPHOST)
+    // Detect if the process has console
+    // Note that this will fail even if a GUI process has STDOUT redirected. So this pretty effectively detects
```

Comment:
Have we lifted this logic from somewhere else? I would love to use the prior art here.

Reply:
I came up with this - well - I read the docs for CreateFile and GetStdHandle and I tested them on both WinExe and Exe in various scenarios. While doing this I did some searching, but could not find anything existing. But I admit I didn't try super hard (I didn't expect this to be a tricky problem, and so far it doesn't seem to be).

Reply:
@vitek-karas If you want me to do this, I'd be happy to...

Reply:
@sdmaclea It would be great if you could also try to search for prior art here. I just did it again, found different answers but none even close to the quality we need (the most common answer is - try Console.ReadLine, if it throws, there's no console... not very good).
In general it seems that the ability to open the console "file" is a good indication, as it's intentionally "special" in Windows. Note that it's designed to "ignore" redirected input/output streams which is what we want here for consistency.
The other alternative I'm aware of is to open the .exe itself and look in the PE header to determine if it's marked GUI or not.

Reply:
There's one scenarios which things might become "interesting" - that is a Windows Service. I guess I'll have to test that case, I don't know if services get console (even hidden) or not. We would want those to behave as console apps though (services can't show UI by default).

Reply:
I also looked for prior art, but like Vitek said it is very immature.  @jkotas how important is using prior art.

I have no idea how to test a windows service.  Maybe this can sit for a while

I am wondering if the right approach might be to open the header and look for the gui bit.  The code must exist somewhere.  It seems like there would be prior art of sorts.

Reply:
We have the code which sets the GUI bit (our SDK does that, I actually wrote that according to the PE spec, it's pretty simple). So if we feel that checking the GUI bit is a better way - we can definitely do that - with one condition - I don't know (yet) how to get to the memory address of a loaded module (we would need to find the .exe module, find the address it's mapped to and be allowed to read through some of its pages to get to the GUI bit).

Reply:
I do not know myself what is better. Thanks for doing the due diligence. The approach looks fine to me given what we have found.

Reply:
I'm trying the service … will report back

Reply:
Windows Service seems to be rather tricky. It behaves as a GUI app (no console or stdout/stderr available), but it also doesn't like message boxes (in fact when I tried it, the service got into a state where it can't be removed by the OS - probably waits for a reboot as it's blocking some process somewhere).

So I'll probably look into the GUI bit reading.

Reply:
> we would need to find the .exe module, find the address it's mapped to

`GetModule(NULL)` does exactly that.

> l that checking the GUI bit is a better

The interesting question is whether it is:
- better to be too aggressive, (display the message box even when it is necessary and a headless app can get stuck because of it)
- or whether it is better to be too conservative and miss displaying the dialog some of the time

I think it is better to be too conservative category, at least to start with. Message boxes on headless are super annoying to workaround. Missing the diagnostic is much easier to workaround.

If we wanted to check the GUI bit, we may want to do it in addition to what is in the change today. It would make us more conservative. (Just exploring the options, I am not saying that we should do it.)

Reply:
Thanks @danmosemsft 
I don't know about .NET Framework, but I kind of doubt it. I'll see if I can maybe hack it somehow to make it fail and test this.

Current plan:
Switch to use the "find the GUI bit in the PE headers" instead of trying to open the console streams. It is less prone to issues caused by some weird setup of the app. Services being one such example.
I do agree that it's better to be conservative and only show the message box when it should work, which should be the case with the GUI bit.

Reply:
Some .NET Framework startup issues of services certainly go into the Windows event log, e.g. unhandled managed exceptions during startup. It's the first place we look at when a service doesn't start up. If its supported to write Windows services in .NET Core this definitely needs to be covered, otherwise diagnosing startup issues is going to be insane for the poor developers diagnosing customer errors. After all if it doesn't start up we can't add custom logging so the runtime needs to have *something* in place to diagnose those startup problems, even if its just some flag we can set telling the runtime to log to a file instead of the console.

Reply:
In 3.0 we've already added the support to redirect all tracing (which includes all errors as well and you can control verbosity) into a file through environment variable. Not sure if this is the best solution for windows services, but at least something.
See #4585 .